tests/integrationv2/test_record_padding.py

```diff
@@ -0,0 +1,136 @@
+        port=port,
+        cipher=cipher,
+        cert=certificate.cert,
+        curve=curve,
+        data_to_send=random_bytes,
+        insecure=True,
+        protocol=protocol,
+        extra_flags=['-record_padding', padding_size]
+    )
+
+    server_options = copy.copy(client_options)
+    server_options.data_to_send = None
+    server_options.mode = Provider.ServerMode
+    server_options.key = certificate.key
+    server_options.extra_flags = None
+
+    s2nd = managed_process(S2N, server_options, timeout=5)
+    openssl = managed_process(provider, client_options, timeout=5)
+
+    # expected length is the padding size + 16 bytes of aead tag
+    expected_total_length = padding_size + 16
+    expected_record_header = get_record_header(expected_total_length)
+
+    for client_results in openssl.get_results():
+        client_results.assert_success()
+        # verify that the openssl is sending padded payloads with the expected size
+        assert to_bytes(expected_record_header) in client_results.stdout
+
+    expected_version = get_expected_s2n_version(protocol, provider)
+
+    for server_results in s2nd.get_results():
+        server_results.assert_success()
+        # verify that the payload was correctly received by the server
+        assert random_bytes in server_results.stdout
+        # verify that the version was correctly negotiated
+        assert to_bytes("Actual protocol version: {}".format(
+            expected_version)) in server_results.stdout
+        # verify that the cipher was correctly negotiated
+        assert to_bytes("Cipher negotiated: {}".format(
+            cipher.name)) in server_results.stdout
+
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [OpenSSL])
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+# only tls 1.3 supports record padding
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", MINIMAL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("padding_size", PADDING_SIZES, ids=get_parameter_name)
+def test_s2n_client_handles_padded_records(managed_process, cipher, provider, curve, protocol, certificate,
+                                           padding_size):
+    port = next(available_ports)
+
+    random_bytes = data_bytes(PAYLOAD_SIZE)
+    server_options = ProviderOptions(
+        mode=Provider.ServerMode,
+        port=port,
+        cipher=cipher,
+        curve=curve,
+        cert=certificate.cert,
+        key=certificate.key,
+        insecure=True,
+        protocol=protocol,
+        extra_flags=[
+            '-record_padding', padding_size]
+    )
+
+    client_options = copy.copy(server_options)
+    client_options.mode = Provider.ClientMode
+    client_options.extra_flags = None
+    client_options.data_to_send = random_bytes
+
```

Comment:
Shouldn't the server, not the client, be sending the records in this test? Aren't we trying to test how the s2n client handles receiving padded records?

And if that's the case here, why is the test still passing if the client doesn't have any knowledge of padding_size?

Reply:
You're right; actually it doesn't matter if the client sends data or not. What we're testing here is that the server sends padded records and that the client is able to process them. Here, the server is simply sending padded handshake messages and we check if openssl is sending padded records as expected and that s2nc is able to process them

Reply:
I'd prefer we test both application data and handshake messages.

Reply:
Added server sending application data + verify client received data